src/AppInstallerCLICore/Workflows/UninstallFlow.cpp

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "UninstallFlow.h"
+#include "WorkflowBase.h"
+#include "MSStoreInstallerHandler.h"
+#include "ShellExecuteInstallerHandler.h"
+#include "AppInstallerMsixInfo.h"
+
+using namespace AppInstaller::Manifest;
+using namespace AppInstaller::Msix;
+using namespace AppInstaller::Repository;
+
+namespace AppInstaller::CLI::Workflow
+{
+    namespace
+    {
+        std::optional<DWORD> InvokeCreateProcess(const std::string& command, IProgressCallback& progress)
+        {
+            std::wstring commandUtf16 = Utility::ConvertToUTF16(command);
+
+            // Parse the command string as application and command line for CreateProcess
+            wil::unique_cotaskmem_string app = nullptr;
+            wil::unique_cotaskmem_string commandLine = nullptr;
+            THROW_IF_FAILED(SHEvaluateSystemCommandTemplate(commandUtf16.c_str(), &app, &commandLine, NULL));
+
+            STARTUPINFOW startupInfo = { sizeof(startupInfo) };
+            wil::unique_process_information processInfo = {};
+            BOOL success = CreateProcessW(app.get(), commandLine.get(), NULL, NULL, FALSE, 0, NULL, NULL, &startupInfo, &processInfo);
+
+            if (!success)
+            {
+                return GetLastError();
+            }
+
+            // Wait for uninstallation to finish
+            while (!progress.IsCancelled())
+            {
+                DWORD waitResult = WaitForSingleObject(processInfo.hProcess, 250);
+                if (waitResult == WAIT_OBJECT_0)
+                {
+                    break;
+                }
+
+                if (waitResult != WAIT_TIMEOUT)
+                {
+                    THROW_LAST_ERROR_MSG("Unexpected WaitForSingleObjectResult: %d", waitResult);
+                }
+            }
+
+            if (progress.IsCancelled())
+            {
+                return {};
+            }
+            else
+            {
+                DWORD exitCode = 0;
+                GetExitCodeProcess(processInfo.hProcess, &exitCode);
+                return exitCode;
+            }
+        }
+    }
+
+    void GetUninstallInfo(Execution::Context& context)
+    {
+        auto installedPackageVersion = context.Get<Execution::Data::InstalledPackageVersion>();
+        const std::string installedTypeString = installedPackageVersion->GetMetadata()[PackageVersionMetadata::InstalledType];
```

Comment:
>[PackageVersionMetadata::InstalledType] [](start = 86, length = 39)

I think John mentioned that this key may not always be present, or this is not the case any more?

Reply:
No, that one should always be present.  It may not be super accurate, but it should at least get the ARP/MSIX split that is needed here.

---
In reply to: [536451002](https://github.com/Microsoft/winget-cli/pull/659#discussion_r536451002) [](ancestors = 536451002)